src/middleware/TestingHandler.ts

```diff
@@ -0,0 +1,279 @@
+				if (testingHandlerOptions.statusCode === undefined) {
+					// manual mode with no status code, can be a global level or request level without statusCode
+					if (requestURL === this.redirectURL) {
+						// we send a 404 after a single redirect if it's done (chain contains redirect Handler)
+						testingHandlerOptions.statusCode = 404;
+					}
+					const relativeURL: string = this.getRelativeURL(requestURL);
+					if (this.manualMap.get(relativeURL) !== undefined) {
+						// checking Manual Map for exact match
+						if (this.manualMap.get(relativeURL).get(requestMethod) !== undefined) {
+							testingHandlerOptions.statusCode = this.manualMap.get(relativeURL).get(requestMethod);
+						} else {
+							throw new Error("API not available in map");
+						}
+					} else {
+						// checking for regex match if exact match doesn't work
+						this.manualMap.forEach((value: Map<string, number>, key: string) => {
+							const regexURL: RegExp = new RegExp(key + "$");
+							if (regexURL.test(relativeURL)) {
+								if (this.manualMap.get(key).get(requestMethod) !== undefined) {
+									testingHandlerOptions.statusCode = this.manualMap.get(key).get(requestMethod);
+								} else {
+									throw new Error("API not available in map");
+								}
+							}
+						});
+					}
+
+					// Case of redirection or request url not in map
+					if (testingHandlerOptions.statusCode === undefined) {
+						throw new Error("API not available in map");
+					}
+				} else {
+					// if we have got redirection
+					if (requestURL === this.redirectURL) {
+						const statusCode: number = testingHandlerOptions.statusCode;
+						if (statusCode === 301 || statusCode === 302 || statusCode === 303 || statusCode === 307 || statusCode === 308) {
+							// we send a 404 after a single redirect if it's done (chain contains redirect Handler)
+							testingHandlerOptions.statusCode = 404;
+						}
+					}
+				}
+			} else {
+				// Handling the case of Random here
+				testingHandlerOptions.statusCode = this.getRandomStatusCode(requestMethod);
+			}
+		} catch (error) {
+			throw error;
+		}
+	}
+
+	/**
+	 * To get the options for execution of the middleware
+	 * @private
+	 * @param {Context} context - The context object
+	 * @returns options for middleware execution
+	 */
+	private getOptions(context: Context): TestingHandlerOptions {
+		let options: TestingHandlerOptions;
+		if (context.middlewareControl instanceof MiddlewareControl) {
+			options = context.middlewareControl.getMiddlewareOptions(TestingHandlerOptions) as TestingHandlerOptions;
+		}
+		if (typeof options === "undefined") {
+			options = Object.assign(new TestingHandlerOptions(), this.options);
+		}
+
+		return options;
+	}
+
+	/**
+	 * To execute the current middleware
+	 * @public
+	 * @async
+	 * @param {Context} context - The context object of the request
+	 * @returns A Promise that resolves to nothing
+	 */
+	public async execute(context: Context): Promise<void> {
```

Comment:
We should not be returning failure conditions for every request.  For random we should only return failures on a small percentage of requests.  That percentage should be configurable.  For manual configuration we should only return errors if the request URL maps to one of the failure cases.

The goal of this handler is that a developer can actually work with the application and can verify that the application can recover from failures.  This would not work if every response was a failure.

Reply:
@darrelmiller 
- Do we need to increase the priority of the success scenarios for the Random case?
- Do we need to give users the option of configuring the type of requests in Random mode?
- For the Manual mode, we return an error only when map points to one of the failure scenarios (we even consider 3xx case as an error scenario, in js-SDK and hence error objects are returned).

Reply:
@krototype We shouldn't return success status codes.  We should pass the request through the testing handler to the Graph API and allow the regular request to pass through.  The idea of the testing handler is to allow a developer to use their application but have random errors be injected into their application so that they can see whether the application handles server generated failures.  I realize now that when we had the conversation about what status code were returned that I got confused sidetracked away from the original goal.  The only status codes that we should return are ones are difficult to manually trigger.  e.g. 5XX and 429.  The rest of the error codes like 4XX and 3XX are unnecessary to ever return in random mode.  I need to investigate more regarding manual mode because I don't see how 3XX could ever be considered an error.